js/util/polygon_poi.js

```diff
@@ -0,0 +1,150 @@
+'use strict';
+
+var isClosedPolygon = require('./util').isClosedPolygon;
+
+function pointToLineDistance(x, y, x1, y1, x2, y2) {
+    var C = x2 - x1;
+    var D = y2 - y1;
+    var lenSq = C * C + D * D;
+    var param = -1;
+    if (lenSq !== 0) {
+        var A = x - x1;
+        var B = y - y1;
+        var dot = A * C + B * D;
+        param = dot / lenSq;
+    }
+    var xx, yy;
+    if (param < 0) {
+        xx = x1;
+        yy = y1;
+    } else if (param > 1) {
+        xx = x2;
+        yy = y2;
+    } else {
+        xx = x1 + param * C;
+        yy = y1 + param * D;
+    }
+    var dx = x - xx;
+    var dy = y - yy;
+    // return Math.sqrt(dx * dx + dy * dy);
+    return (dx * dx + dy * dy);
+}
+
+function pointToPerimeterDistance(x, y, points) {
+    var d, p1, p2, minDistance = Number.MAX_VALUE;
+    var len = points.length;
+
+    for (var i = -1, l = len, j = l - 1; ++i < l; j = i) {
+        p1 = points[i];
+        p2 = points[j];
+        d = pointToLineDistance(x, y, p1.x, p1.y, p2.x, p2.y);
+        if (d < minDistance) {
+            minDistance = d;
+        }
+    }
+    return minDistance;
+}
+
+function isInside(x, y, points) {
+    var c = false;
+    for (var i = -1, l = points.length, j = l - 1; ++i < l; j = i) {
+        var pi = points[i];
+        var pj = points[j];
+        if (((pi.y <= y && y < pj.y) || (pj.y <= y && y < pi.y)) &&
+            (x < (pj.x - pi.x) * (y - pi.y) / (pj.y - pi.y) + pi.x))
+            c = !c;
+    }
+    return c;
+}
```

Comment:
Many of the functions introduced here (as well as in `util.js`) are also present in https://github.com/mapbox/mapbox-gl-js/blob/62ac39cf7f1f55a9a82b44e67fb5d71271e97f8b/js/util/intersection_tests.js#L102-L164. Can we eliminate the duplication?